Add betrayal scenario support for original campaign

ihhub opened the pull request:
This pull request fixes multiple issues related to campaigns:
- fix multiple warnings / code smells detected by static analysers
- fix logic of detecting last scenario of a campaign ( #4985 ), especially `isLastScenario` method. Also scenario sequence was incorrect
- add an ability of betraying a lord in the original campaign. Previously we were working only with scenario ID assuming that we can't change campaigns. In the proposed code we save scenario ID and campaign ID for each scenario. This allows us to track which scenarios are completed and which will lead to a different campaign
- add a popup confirmation window for scenario restart
- some code cleanup and optimization

close #3491
close #4985

a1exsh commented:
@ihhub you beat me to it! :D

oleg-derevenetz reviewed (commented):
Hi @ihhub no obvious issues caught my attention, except one minor note. Please take a look when it's convenient for you.

ihhub commented on src/fheroes2/campaign/campaign_scenariodata.h:
```diff
@@ -50,6 +50,33 @@ namespace Campaign
+        ScenarioInfoId() = default;
+
+        // Make sure that we don't pass just one argument when 2 are required.
+        ScenarioInfoId( const int ) = delete;
```
I am not sure why it was working but somehow during the changes 1 argument was acceptable as well. Now it is not. I removed it.

ihhub commented:
Hi @Branikolog , could you please test these changes as the highest priority?

a1exsh commented:
@ihhub cool, seems like it just works! :D

I gave it a try and I notice that the audio sequence of Roland addressing the player with "Come and fight on my side" is missing in the Archibald's campaign.

The Archibald addressing the player in Roland's campaign is present, but if the player hits a key early to skip intro he also skips the 2nd sequence.  This seems to be the OG problem, actually, i.e. in one campaign there are two separate sequences, but in the other — just one.

Branikolog commented:
Hello, @ihhub 
I've noticed next thing:
In the betrayal scenario we don't have previously earned rewards. (they are visible at scenario selection window, but not applied after start)
OG:
![image](https://user-images.githubusercontent.com/55348946/152675820-f328d13d-90e4-44f0-96e4-7a0d528dcecb.png)
fheroes2:
![image](https://user-images.githubusercontent.com/55348946/152676049-421a07ca-a6a6-4742-8d67-b09b235f5437.png)
The next scenario this reward is still applied too.
OG has none:
![image](https://user-images.githubusercontent.com/55348946/152676226-44498539-67cb-4489-bd93-19b19600154d.png)
fheroes2:
![image](https://user-images.githubusercontent.com/55348946/152676237-1bff732a-ff35-4576-b349-751d3c93968b.png)

Also, I've noticed, that RESTART button doesn't perform confirmation dialog.
OG:
![image](https://user-images.githubusercontent.com/55348946/152676086-b6c0a735-16e6-4265-b50a-32d27150c713.png)

ihhub commented:
Hi @a1exsh and @Branikolog , thanks for your findings! I will fix them soon.

oleg-derevenetz commented:
@ihhub how about we hold off on the new release for a week to include this (well-tested) PR in it?

ihhub commented:
> @ihhub how about we hold off on the new release (0.9.12) for a week to include this (well-tested) PR in it?

I've marked it for the next release :)

oleg-derevenetz commented:
@ihhub

> I've marked it for the next 0.9.13 release :)

Yeah, that's why I asking :) This is a good change and it's not fair that people will only get it in a month :)

ihhub commented:
> @ihhub
> 
> > I've marked it for the next 0.9.13 release :)
> 
> Yeah, that's why I'm asking :) This is a good change and it's not fair that people will only get it in a month :)

Let me try then to fix the rest of issues fast. I'll do it today and we'll test again.

ihhub commented:
Hi @a1exsh and @Branikolog , I addressed all your comments. Could you please take a look again?

@Branikolog , also please verify that other campaigns work properly and old saves are still in working state.

oleg-derevenetz reviewed (changes requested):
Hi @ihhub I just noticed a few TYPOS, nothing serious.

a1exsh commented:
@ihhub the missing video sequence is in place now.  Great piece of work!

ihhub commented:
> Hi @ihhub I just noticed a few typos, nothing serious.

Thank you for the review! I've fixed the typos.
